sdb/commands/zfs/dbuf.py

```diff
@@ -0,0 +1,123 @@
+#
+# Copyright 2019 Delphix
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+# pylint: disable=missing-docstring
+
+import argparse
+from typing import Iterable
+
+import drgn
+import sdb
+
+
+class DslDir():
+
+    @staticmethod
+    def name(dd):
+        pname = ''
+        if dd.dd_parent:
+            pname = DslDir.name(dd.dd_parent) + '/'
+        return pname + dd.dd_myname.string_().decode("utf-8")
+
+
+class Dataset():
+
+    @staticmethod
+    def name(ds):
+        if not ds:
+            return 'MOS'
+        name = DslDir.name(ds.ds_dir)
+        if not ds.ds_prev:
+            sn = ds.ds_snapname.string_().decode("utf-8")
+            if len(sn) == 0:
+                sn = '%UNKNOWN_SNAP_NAME%'
+            name += '@' + sn
+        return name
+
+
+class Objset():
+
+    @staticmethod
+    def name(os):
+        return Dataset.name(os.os_dsl_dataset)
+
+
+class Dbuf(sdb.Locator, sdb.PrettyPrinter):
+
+    names = ["dbuf"]
+    input_type = "dmu_buf_impl_t *"
+    output_type = "dmu_buf_impl_t *"
+
+    def _init_argparse(self, parser: argparse.ArgumentParser) -> None:
+        parser.add_argument(
+            '-o',
+            '--object',
+            type=int,
+            help='filter: only dbufs of this object')
+        parser.add_argument(
+            '-l',
+            '--level',
+            type=int,
+            help='filter: only dbufs of this level')
+        parser.add_argument(
+            '-b',
+            '--blkid',
+            type=int,
+            help='filter: only dbufs of this blkid')
+        parser.add_argument(
+            '-d',
+            '--dataset',
+            type=str,
+            help='filter: only dbufs of this dataset name (or "MOS")')
+        parser.add_argument(
+            '-H',
+            '--has-holds',
+            action='store_true',
+            help='filter: only dbufs that have nonzero holds')
+
+    def pretty_print(self, dbufs):
+        print("{:>20} {:>8} {:>4} {:>8} {:>5} {}".format(
+            "addr", "object", "lvl", "blkid", "holds", "os"))
+        for dbuf in dbufs:
+            print("{:>20} {:>8d} {:>4d} {:>8d} {:>5d} {}".format(
+                hex(dbuf),
+                int(dbuf.db.db_object),
+                int(dbuf.db_level),
+                int(dbuf.db_blkid),
+                int(dbuf.db_holds.rc_count),
+                Objset.name(dbuf.db_objset)))
```

Comment:
One thing to think about this now or in the future would be to potentially use the `Table` class to print the above.
You are probably the best to judge if this really worths it though as you are the heaviest user of this command. Some questions I'd ask is:
1) Do you think you would ever want to  print something else besides the above (by default or optionally?)
2) Would you ever to have the above output sorted in any way?

Again, what you have looks good I just thought I'd bring it up.

Reply:
The one big downside of the Table (IIRC) is that you have to gather all the info before it prints anything.  In my testing on a very small system, it takes minutes for `dbuf` to run to completion so I've almost always `^C`'ed out of it while testing.  Even with `mdb`, `::dbufs` could take a long time to run.  And it's really nice to get incremental output for long-running commands.

Reply:
Makes sense, thanks for pointing that out!